clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2123,6 +2123,71 @@ namespace {
       }
    }
 
+   void ResetKilledBounds(StmtDeclSetTy &KilledBounds,
+                          BoundsContextTy &BoundsCtx,
+                          Stmt *S) {
+     auto I = KilledBounds.find(S);
+     if (I == KilledBounds.end())
+       return;
+
+     for (const VarDecl *V : I->second) {
+       VarDecl *VD = const_cast<VarDecl *>(V);
+       BoundsExpr *Bounds = VD->getBoundsExpr();
+       assert(Bounds && "Invalid bounds");
+
+       if (isa<CountBoundsExpr>(Bounds))
+         Bounds = ExpandToRange(VD, Bounds);
+       BoundsCtx[VD] = Bounds;
+     }
+   }
+
+   void UpdateBoundsInContext(BoundsMapTy &WidenedBounds,
+                              BoundsContextTy &BoundsCtx) {
+     for (const auto item : WidenedBounds) {
+       VarDecl *V = const_cast<VarDecl *>(item.first);
+       unsigned Offset = item.second;
+
+       BoundsExpr *Bounds = V->getBoundsExpr();
+       if (!Bounds)
+         continue;
+
+       // If the declared bounds are CountBoundsExpr, normalize them to
+       // RangeBoundsExpr.
+       if (isa<CountBoundsExpr>(Bounds)) {
+         if (!V->getType()->isArrayType())
+           Bounds = ExpandToRange(V, Bounds);
+         else {
+           ExprResult ER = S.BuildDeclRefExpr(V, V->getType(),
+                                              clang::ExprValueKind::VK_LValue,
+                                              SourceLocation());
+           if (ER.isInvalid())
+             continue;
+           Expr *Base = ER.get();
+           Bounds = ArrayExprBounds(Base);
+         }
+       }
+
+       RangeBoundsExpr *RBE = dyn_cast<RangeBoundsExpr>(Bounds);
+       assert(RBE && "Invalid bounds");
+
+       const llvm::APInt
+         APIntOff(Context.getTargetInfo().getPointerWidth(0), Offset);
+       IntegerLiteral *WidenedOffset = CreateIntegerLiteral(APIntOff);
+
+       Expr *Lower = RBE->getLowerExpr();
+       Expr *Upper = RBE->getUpperExpr();
+
+       // NewUpperBound = UpperBound + WidenedOffset.
+       Expr *NewUpper = ExprCreatorUtil::CreateBinaryOperator(
```

Comment:
I tested this locally and ran into some issues with checking an assignment using widened bounds.

In this example:
```
void widen_bounds_bug(nt_array_ptr<int> a : bounds(a, a + 1),
                      nt_array_ptr<int> p : bounds(p, p + 0),
                      nt_array_ptr<int> b : bounds(b, b + (0 + 1))) {
  if (*p) {
    // Bounds of p have been widened to (p, (p + 0) + 1).
    // The target bounds of a are (a, a + 1).
    // Because the base of p's bounds is p + 0,
    // ProveBoundsDeclValidity is unable to create a base range for p's bounds,
    // and is therefore unable to prove the validity of a's bounds after the assignment.
    a = p;

    // Bounds of b are (b, b + (0 + 1)).
    // The target bounds of a are (a, a + 1).
    // Because the base of b's bounds is b,
    // ProveBoundsDeclValidity is able to create a base range for b's bounds,
    // and is therefore able to prove the validity of a's bounds after the assignment.
    a = b;
  }
}
```
The assignment `a = p` results in a warning "unable to prove the declared bounds of a are valid after the assignment". The reason is that the new upper expression for p's widened bounds `(p, (p + 0) + 1)` is left-associative, so the base of the widened bounds is `p + 0`. If the upper expression were right-associative `(p, p + (0 + 1))`, as in the declared bounds for `b`, then the assignment `a = p` would be fine (as in the assignment `a = b`).

I quickly put together a local fix involving `SplitIntoBaseAndOffset` to split the upper expression of a variable's bounds and use the resulting base and offset to create the new upper expression for the widened bounds, but I haven't fully tested it. Below is the rough approach:
```
Expr *UpperBase;
llvm::APSInt  UpperOffsetConstant(1, true);
Expr *UpperOffsetVariable = nullptr;
SplitIntoBaseAndOffset(Upper, UpperBase, UpperOffsetConstant, UpperOffsetVariable);

Expr *WidenedUpper;
if (UpperOffsetVariable) {
  WidenedUpper = ExprCreatorUtil::CreateBinaryOperator(S, UpperOffsetVariable, WidenedOffset, BinaryOperatorKind::BO_Add);
} else {
  IntegerLiteral *UpperConst = CreateIntegerLiteral(UpperOffsetConstant);
  WidenedUpper = ExprCreatorUtil::CreateBinaryOperator(S, UpperConst, WidenedOffset, BinaryOperatorKind::BO_Add);
}

Expr *NewUpper = ExprCreatorUtil::CreateBinaryOperator(S, UpperBase, WidenedUpper, BinaryOperatorKind::BO_Add);
```

Reply:
Yes, SplitIntoBaseAndOffset does currently handle associativity and multiple arithmetic operators correctly. In BoundsAnalysis I ran into a similar issue wherein I had to correctly determine the offset of the expression being dereferenced. So I wrote an implementation of SplitIntoBaseAndOffset (see https://github.com/microsoft/checkedc-clang/blob/master/clang/lib/Sema/BoundsAnalysis.cpp#L340). This function associates all variables to the LHS and all constants to the RHS. Ideally, we want to merge both these functions into one.

One thing to keep in mind with changing associativity of operators is overflow. The original expression may not overflow but if we change the associativity then an overflow/underflow may occur.